brainscore_language/model_helpers/container.py

```diff
@@ -0,0 +1,319 @@
+import functools
+import json
+import logging
+import multiprocessing
+import re
+import subprocess
+import sys
+from collections import OrderedDict
+from pathlib import Path
+from typing import List, Tuple, Dict, Union, Callable
+
+import numpy as np
+import torch
+import xarray as xr
+from joblib import Parallel, delayed, parallel_backend
+from numpy.core import defchararray
+from tqdm import tqdm
+
+from brainio.assemblies import DataAssembly, NeuroidAssembly, BehavioralAssembly
+from brainscore_language.artificial_subject import ArtificialSubject
+from brainscore_language.utils import fullname
+
+
+class ContainerSubject(ArtificialSubject):
+    """
+    Evaluation interface for arbitary containerized models.
+    User must install either 'Singularity' or 'Docker' to evaluate container models.
+
+    To add new model, build a container with an entry point that supports the following interface:
+
+    $CONTAINER_BACKEND run $CONTAINER_NAME $ENTRYPOINT \
+    --model <model_identifier> --measure <measure> --context <context> --text <text>
+
+    The entrypoint must return a JSON string to stdout with the following format based on the following tasks:
+
+    Task: predict next word
+    Input: --measure "next-word"
+    Output: {"measure": NEXT_WORD} 
+    where NEXT_WORD is a string
+
+    Task: estimate reading times
+    Input: --measure "token-logits"
+    Output: {"tokens": TOKENS, "measure": LOGITS} 
+    where LOGITS are the logits predicting the target tokens and TOKENS are the corresponding true token indices
+    LOGITS are thus generated using the text up until the last token whereas TOKENS are the true tokens
+    Both LOGITS and TOKENS should have the same shape[0], and shape[1] should be the vocabulary size for the LOGITS
+    Objects should be returned as lists, e.g., object.cpu().numpy().tolist() for pytorch tensors
+
+    Task: extract representation
+    Input: --measure MEASURE
+    Output: {"measure": REPRESENTATION}
+    where MEASURE is the name of the representation as supported by your container 
+    and REPRESENTATION is an array of shape (1, representation_size) cast to a list
+
+    Note: While the internals of any containerized model are not restricted, the interface must be as described above. 
+    It is highly recommended to raise detailed error messages from inside the container, so they can be escalated here.
+    It is also recommended to include a list of supported measures in the container's documentation.
+    """
```

Comment:
love it. Maybe you could add a small blurb on how to add new measures, e.g. when we define a new task (a near-term one is likely going to be next-word surprisal)

Reply:
added a paragraph will commit once I finish these comments.